sdk/nodejs/npm/manager_test.go

```diff
@@ -51,31 +51,146 @@ func chdir(t *testing.T, dir string) {
 	})
 }
 
-func testInstall(t *testing.T, expectedBin string, production bool) {
+func TestResolvePackageManager(t *testing.T) {
+	t.Parallel()
+	for _, tt := range []struct {
+		name      string
+		exepcted  string
+		lockFiles []string
+	}{
+		{"defaults to npm", "npm", []string{}},
+		{"picks npm", "npm", []string{"npm"}},
+		{"picks yarn", "yarn", []string{"yarn"}},
+		{"picks pnpm", "pnpm", []string{"pnpm"}},
+		{"yarn > pnpm > npm", "yarn", []string{"yarn", "pnpm", "npm"}},
+		{"pnpm > npm", "pnpm", []string{"pnpm", "npm"}},
+	} {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			dir := t.TempDir()
+			for _, lockFile := range tt.lockFiles {
+				writeLockFile(t, dir, lockFile)
+			}
+			pm, err := ResolvePackageManager(dir)
+			require.NoError(t, err)
+			require.Equal(t, tt.exepcted, pm.Name())
+		})
+	}
+}
+
+func TestPack(t *testing.T) {
+	t.Parallel()
+
+	packageJSON := []byte(`{
+	    "name": "test-package",
+		"version": "1.0"
+	}`)
+
+	for _, pm := range []string{"npm", "yarn", "pnpm"} {
+		pm := pm
+		t.Run(pm, func(t *testing.T) {
+			t.Parallel()
+			dir := t.TempDir()
+			writeLockFile(t, dir, pm)
+			packageJSONFilename := filepath.Join(dir, "package.json")
+			require.NoError(t, os.WriteFile(packageJSONFilename, packageJSON, 0o600))
+			stderr := new(bytes.Buffer)
+
+			artifact, err := Pack(context.Background(), dir, stderr)
+
+			require.NoError(t, err)
+			// check that the artifact contains a package.json
+			b, err := gzip.NewReader(bytes.NewReader((artifact)))
+			require.NoError(t, err)
+			tr := tar.NewReader(b)
+			for {
+				h, err := tr.Next()
+				if err == io.EOF {
+					require.Fail(t, "package.json not found")
+					break
+				}
+				require.NoError(t, err)
+				if h.Name == "package/package.json" {
+					break
+				}
+			}
+		})
+	}
+}
+
+func TestPackInvalidPackageJSON(t *testing.T) {
+	t.Parallel()
+
+	// Missing a version field
+	packageJSON := []byte(`{
+	    "name": "test-package"
+	}`)
+
+	for _, pm := range []string{"npm", "yarn", "pnpm"} {
+		pm := pm
+		t.Run(pm, func(t *testing.T) {
+			t.Parallel()
+			dir := t.TempDir()
+			writeLockFile(t, dir, pm)
+			packageJSONFilename := filepath.Join(dir, "package.json")
+			require.NoError(t, os.WriteFile(packageJSONFilename, packageJSON, 0o600))
+			stderr := new(bytes.Buffer)
+
+			_, err := Pack(context.Background(), dir, stderr)
+
+			require.Error(t, err)
+		})
+	}
+}
+
+// writeLockFile writes a mock lockfile for the selected package manager
+func writeLockFile(t *testing.T, dir string, packageManager string) {
+	t.Helper()
+	switch packageManager {
+	case "npm":
+		writeFile(t, filepath.Join(dir, "package-lock.json"), "{\"lockfileVersion\": 2}")
+	case "yarn":
+		writeFile(t, filepath.Join(dir, "yarn.lock"), "# yarn lockfile v1")
+	case "pnpm":
+		writeFile(t, filepath.Join(dir, "pnpm-lock.yaml"), "lockfileVersion: '6.0'")
+	}
+}
+
+func testInstall(t *testing.T, expectedBin string, production bool, packageManager string) {
```

Comment:
`expectedBin` and `packageManager` always seem to be the same here. Should we just rename `expectedBin` to `packageManager` instead of passing it in twice?

Reply:
Good idea, thanks.